src/goto-instrument/contracts/contracts.cpp

```diff
@@ -171,33 +207,54 @@ void code_contractst::check_apply_loop_contracts(
       new_tmp_symbol(
         clause.type(), loop_head->source_location(), mode, symbol_table)
         .symbol_expr();
-    havoc_code.add(goto_programt::make_decl(
+    generated_code.add(goto_programt::make_decl(
       old_decreases_var, loop_head->source_location()));
     old_decreases_vars.push_back(old_decreases_var);
 
     const auto new_decreases_var =
       new_tmp_symbol(
         clause.type(), loop_head->source_location(), mode, symbol_table)
         .symbol_expr();
-    havoc_code.add(goto_programt::make_decl(
+    generated_code.add(goto_programt::make_decl(
       new_decreases_var, loop_head->source_location()));
     new_decreases_vars.push_back(new_decreases_var);
   }
 
-  // non-deterministically skip the loop if it is a do-while loop
+  // non-deterministically skip the loop if it is a do-while loop.
+  // TODO: Fix loop contract issues for do/while loops.
   if(!loop_head->is_goto())
   {
-    havoc_code.add(goto_programt::make_goto(
+    generated_code.add(goto_programt::make_goto(
       loop_end,
       side_effect_expr_nondett(bool_typet(), loop_head->source_location())));
   }
 
-  // Now havoc at the loop head.
+  // Assume invariant & decl the variant temporaries (just before loop guard).
   // Use insert_before_swap to preserve jumps to loop head.
-  insert_before_swap_and_advance(goto_function.body, loop_head, havoc_code);
+  insert_before_swap_and_advance(goto_function.body, loop_head, generated_code);
+
+  // Forward the loop_head iterator until the start of the body.
+  // This is necessary because complex C loop_head conditions are sometimes
+  // converted to multiple GOTO instructions (e.g. temporaries are introduced).
+  const auto head_loc = loop_head->source_location();
+  while(loop_head->source_location() == head_loc)
+    loop_head++;
```

Comment:
So we had similar code in the preceding commit. Perhaps factor out this advancing-the-loop-head into a function. Also, I have no idea _why_ it is necessary to forward the `loop_head` iterator.

Reply:
I have added a regression test + more description there and also to the commit message.